packages/library-base/src/commonMain/kotlin/io/realm/internal/query/ObjectQuery.kt

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal.query
+
+import io.realm.RealmObject
+import io.realm.RealmResults
+import io.realm.internal.Mediator
+import io.realm.internal.RealmReference
+import io.realm.internal.RealmResultsImpl
+import io.realm.internal.Thawable
+import io.realm.internal.genericRealmCoreExceptionHandler
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.RealmCoreException
+import io.realm.internal.interop.RealmCoreIndexOutOfBoundsException
+import io.realm.internal.interop.RealmCoreInvalidQueryException
+import io.realm.internal.interop.RealmCoreInvalidQueryStringException
+import io.realm.internal.interop.RealmInterop
+import io.realm.query.RealmQuery
+import io.realm.query.RealmScalarNullableQuery
+import io.realm.query.RealmScalarQuery
+import io.realm.query.RealmSingleQuery
+import io.realm.query.Sort
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+@Suppress("SpreadOperator")
+internal class ObjectQuery<E : RealmObject> constructor(
+    private val realmReference: RealmReference,
+    private val clazz: KClass<E>,
+    private val mediator: Mediator,
+    composedQueryPointer: NativePointer? = null,
+    private val filter: String,
+    private vararg val args: Any?
+) : RealmQuery<E>, Thawable<RealmResultsImpl<E>> {
+
+    private val queryPointer: NativePointer = when {
+        composedQueryPointer != null -> composedQueryPointer
+        else -> parseQuery()
+    }
+
+    private val resultsPointer: NativePointer by lazy {
+        RealmInterop.realm_query_find_all(queryPointer)
+    }
+
+    constructor(
+        composedQueryPointer: NativePointer?,
+        objectQuery: ObjectQuery<E>
+    ) : this(
+        objectQuery.realmReference,
+        objectQuery.clazz,
+        objectQuery.mediator,
+        composedQueryPointer,
+        objectQuery.filter,
+        *objectQuery.args
+    )
+
+    override fun find(): RealmResults<E> =
+        RealmResultsImpl(realmReference, resultsPointer, clazz, mediator)
+
+    override fun query(filter: String, vararg arguments: Any?): RealmQuery<E> {
+        val appendedQuery = tryCatchCoreException {
+            RealmInterop.realm_query_append_query(queryPointer, filter, *arguments)
+        }
+        return ObjectQuery(appendedQuery, this)
+    }
+
+    // TODO Descriptors are added using 'append_query', which requires an actual predicate. This
```

Comment:
An issue would help tracking this task.

Reply:
Just an optimization, so just added the `OPTIMIZE` tag to the TODO.